sql: panic: internal error: indexed var must be type checked first

This is both an internal error and a panic. Cool.

```
WITH
    with_20394 (col_162526)
        AS (
            SELECT * FROM (VALUES ('d2d225e2-e9be-4420-a645-d1b8f577511c':::UUID)) AS tab_25520 (col_162526)
            UNION ALL
                SELECT
                    *
                FROM
                    (
                        VALUES
                            ('1d6eaf81-8a2c-43c5-a495-a3b102917ab1':::UUID),
                            (NULL),
                            ('01ca0d62-8ae4-41e8-a2dd-06b8f1a41e44':::UUID),
                            ('23c786e5-3de2-45b3-98d0-c8e2377f2389':::UUID),
                            ('7f75f876-6f5c-4368-8f52-d942c45958f2':::UUID),
                            (NULL)
                    )
                        AS tab_25521 (col_162527)
        ),
    with_20395 (col_162528, col_162529, col_162530)
        AS (
            SELECT
                *
            FROM
                (
                    VALUES
                        (0.0959175465027563:::FLOAT8, NULL, NULL),
                        (0.34630617920365453:::FLOAT8, 4427567434424110975:::INT8, 7285865664883706941:::INT8),
                        (NULL, (-6148712980915792398):::INT8, 5369844642347829590:::INT8),
                        (0.06208650787628556:::FLOAT8, (-2387525628515147251):::INT8, 469309070992713473:::INT8)
                )
                    AS tab_25522 (col_162528, col_162529, col_162530)
        )
SELECT
    date_trunc(e'\x1aF':::STRING::STRING, '1970-01-01 05:58:56.000141+00:00':::TIMESTAMP::TIMESTAMP)::TIMESTAMP
        AS col_162531,
    B'10110111010001110110001101010101010000111000110011110100100101111101100010111000111000101010101101'
        AS col_162532,
    current_database()::STRING AS col_162533,
    max(with_20394.col_162526::UUID)::UUID AS col_162534,
    sqrt((-33600271530240.79860):::DECIMAL::DECIMAL)::DECIMAL AS col_162535,
    '23:54:57.078405':::TIME AS col_162536,
    isnan(0.8353995614887574:::FLOAT8::FLOAT8)::BOOL AS col_162537,
    date_trunc('_x':::STRING::STRING, '1970-01-01 07:14:26.000689+00:00':::TIMESTAMPTZ::TIMESTAMPTZ)::TIMESTAMPTZ
        AS col_162538,
    statement_timestamp():::TIMESTAMPTZ::TIMESTAMPTZ AS col_162539,
    pi()::FLOAT8 AS col_162540,
    3697877132:::OID AS col_162541,
    convert_to(NULL::STRING, e'?,]\x7fWN':::STRING::STRING)::BYTES AS col_162542
FROM
    with_20394
GROUP BY
    with_20394.col_162526
HAVING
    isnan(8.811140830642087955E+27:::DECIMAL::DECIMAL)::BOOL
ORDER BY
    with_20394.col_162526 ASC;
```

```
panic: internal error: indexed var must be type checked first [recovered]
	panic: panic while executing 1 statements: WITH _ (_) AS (SELECT * FROM (VALUES (_:::UUID)) AS _ (_) UNION ALL SELECT * FROM (VALUES (_:::UUID), (__more5__)) AS _ (_)) , _ (_, _, _) AS (SELECT * FROM (VALUES (_:::FLOAT8, _, _), (__more3__)) AS _ (_, _, _)) SELECT date_trunc(_::STRING, _::TIMESTAMP)::TIMESTAMP AS _, _ AS _, current_database()::STRING AS _, max(_._::UUID)::UUID AS _, sqrt((_):::DECIMAL::DECIMAL)::DECIMAL AS _, _:::TIME AS _, isnan(_::FLOAT8)::BOOL AS _, date_trunc(_::STRING, _::TIMESTAMPTZ)::TIMESTAMPTZ AS _, statement_timestamp()::TIMESTAMPTZ AS _, pi()::FLOAT8 AS _, _:::OID AS _, convert_to(_::STRING, _::STRING)::BYTES AS _ FROM _ GROUP BY _._ HAVING isnan(_::DECIMAL)::BOOL ORDER BY _._ ASC; caused by internal error: indexed var must be type checked first

goroutine 280 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000a14000, 0x399f120, 0xc0015ac940, 0x315fd40, 0xc00155a080)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:723 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000a14000, 0x399f120, 0xc0015ac940)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:435 +0x61
panic(0x315fd40, 0xc00155a080)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IndexedVar).ResolvedType(0xc001cf3380, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/indexed_vars.go:86 +0x87
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).Init(0xc000cf3418, 0xc0000dcc80, 0xc001090d20, 0x2, 0x2, 0xc0017956c0, 0x3990520, 0xc000ced180, 0xc0009da601, 0xc0017956c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:155 +0x87e
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).InitWithEvalCtx(0xc000cf3400, 0x39ab660, 0xc000cf3400, 0xc0000dcc80, 0xc001090d20, 0x2, 0x2, 0xc001612b40, 0xc0017956c0, 0x4, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:853 +0x135
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Init(0xc000cf3400, 0x39ab660, 0xc000cf3400, 0xc0000dcc80, 0xc001090d20, 0x2, 0x2, 0xc001612b40, 0x4, 0x3990520, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:829 +0x156
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*aggregatorBase).init(0xc000cf3400, 0x39ab660, 0xc000cf3400, 0xc001612b40, 0x4, 0xc0016c2960, 0x39ab6e0, 0xc000ca1b80, 0xc0000dcc80, 0x3990520, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:240 +0xd14
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.newOrderedAggregator(0xc001612b40, 0x4, 0xc0016c2960, 0x39ab6e0, 0xc000ca1b80, 0xc0000dcc80, 0x3990520, 0xc000ced180, 0xc00151dc00, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:378 +0xf8
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.newAggregator(0xc001612b40, 0x4, 0xc0016c2960, 0x39ab6e0, 0xc000ca1b80, 0xc0000dcc80, 0x3990520, 0xc000ced180, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:344 +0x265
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.newProcessor(0x399f1e0, 0xc001cf3440, 0xc001612b40, 0x4, 0xc0000dcbb8, 0xc0000dcc80, 0xc0009da490, 0x1, 0x1, 0xc00172f488, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:1035 +0x21fe
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).makeProcessor(0xc001612b40, 0x399f1e0, 0xc001cf3440, 0xc0000dcba0, 0xc0009da490, 0x1, 0x1, 0x3973160, 0xc00150ef00, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:350 +0x1d5
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupProcessors(0xc001612b40, 0x399f1e0, 0xc001cf3440, 0xc0016b1340, 0x7, 0x7, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:425 +0x1cb
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setup(0xc001612b40, 0x399f1e0, 0xc001cf3440, 0xc001090790, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:525 +0xda
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).setupFlow(0xc000114f00, 0x399f1e0, 0xc001cf2120, 0x39d55a0, 0xc00034d940, 0xc0015c4640, 0xc001090780, 0x398e520, 0xc0017c22c0, 0xc000a14410, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:463 +0x82c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).SetupLocalSyncFlow(0xc000114f00, 0x399f1e0, 0xc001cf2120, 0xc0015c4640, 0xc001090780, 0x398e520, 0xc0017c22c0, 0xc000a14410, 0x1, 0xc000428ae0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:518 +0xe7
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc00049c500, 0xc0016c2780, 0xc000fce630, 0xc001730860, 0xc0017c22c0, 0xc000a14410, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:241 +0x830
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc00049c500, 0x399f1e0, 0xc001cf2120, 0xc000a14410, 0xc0016c2780, 0xc000fce630, 0x39a0e60, 0xc00166a360, 0xc0017c22c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:840 +0x227
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000a14000, 0x399f1e0, 0xc001cf2120, 0xc000a14350, 0x3, 0x7f2ec916f8b8, 0xc0016b0fd0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1121 +0x283
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000a14000, 0x399f1e0, 0xc001cf2120, 0xc000a14350, 0x7f2ec916f8b8, 0xc0016b0fd0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:957 +0x658
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000a14000, 0x399f1e0, 0xc001cf2120, 0x39a4e60, 0xc000bf3e00, 0xc00161b000, 0x7bb, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459 +0xe03
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000a14000, 0x399f1e0, 0xc001cf2120, 0x39a4e60, 0xc000bf3e00, 0xc00161b000, 0x7bb, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101 +0x610
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000a14000, 0x399f1e0, 0xc001cf20f0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1221 +0x1acf
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000a14000, 0x399f120, 0xc0015ac940, 0xc000232e78, 0x5400, 0x15000, 0xc000232f10, 0xc000394ed0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1157 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000c63080, 0x399f120, 0xc0015ac940, 0xc000a14000, 0x5400, 0x15000, 0xc000232f10, 0xc000394ed0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:437 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0002b0129, 0xc000388ca0, 0x399f120, 0xc0015ac940, 0xc000394ed0, 0xc000c63080, 0xc000171b00, 0x39a2e60, 0xc000388c80, 0xc0011fed20, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:581 +0x21f
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:516 +0x17b
```

--- a/pkg/sql/distsqlplan/physical_plan.go
+++ b/pkg/sql/distsqlplan/physical_plan.go
@@ -512,7 +512,10 @@ func (p *PhysicalPlan) AddRendering(
 				if post.Projection {
 					internalColIdx = post.OutputColumns[internalColIdx]
 				}
-				newExpr, err := MakeExpression(&tree.IndexedVar{Idx: int(internalColIdx)}, exprCtx, nil)
+				newExpr, err := MakeExpression(tree.NewTypedOrdinalReference(
+					int(internalColIdx),
+					p.ResultTypes[c.ColIdx].ToDatumType()),
+					exprCtx, nil /* indexVarMap */)
 				if err != nil {
 					return err
 				}

--- a/pkg/sql/logictest/testdata/logic_test/sqlsmith
+++ b/pkg/sql/logictest/testdata/logic_test/sqlsmith
@@ -29,3 +29,43 @@ statement ok
 INSERT INTO table8 (col1, col2, col4, col6)
 VALUES ('19:06:18.321589', NULL, NULL, NULL)
 UNION ALL (SELECT NULL, NULL, NULL, NULL FROM table5 AS tab_8);
+
+# Regression: #36441 (raw indexed var can't be type checked)
+query TO
+WITH
+    with_20394 (col_162526)
+        AS (
+            SELECT
+                *
+            FROM
+                (
+                    VALUES
+                        (
+                            'd2d225e2-e9be-4420-a645-d1b8f577511c':::UUID
+                        )
+                )
+                    AS tab_25520 (col_162526)
+            UNION ALL
+                SELECT
+                    *
+                FROM
+                    (
+                        VALUES
+                            (
+                                '1d6eaf81-8a2c-43c5-a495-a3b102917ab1':::UUID
+                            )
+                    )
+                        AS tab_25521 (col_162527)
+        )
+SELECT
+    max(with_20394.col_162526::UUID)::UUID AS col_162534,
+    3697877132:::OID AS col_162541
+FROM
+    with_20394
+GROUP BY
+    with_20394.col_162526
+ORDER BY
+    with_20394.col_162526 ASC
+----
+1d6eaf81-8a2c-43c5-a495-a3b102917ab1  3697877132
+d2d225e2-e9be-4420-a645-d1b8f577511c  3697877132

